core/services/llo/channeldefinitions/onchain_channel_definition_cache.go

```diff
@@ -228,181 +376,306 @@ func (c *channelDefinitionCache) readLogs(ctx context.Context) (err error) {
+			if err != nil {
+				// Log warning but continue processing other logs
+				c.lggr.Warnw("Failed to unpack adder log", "err", err, "blockNumber", log.BlockNumber)
+				continue
+			}
+			trigger = types.Trigger{
+				Source:   unpacked.ChannelAdderId,
+				URL:      unpacked.Url,
+				SHA:      unpacked.Sha,
+				LogIndex: log.LogIndex,
+				BlockNum: blockNumFromUint64(unpacked.Raw.BlockNumber),
+				TxHash:   log.TxHash,
+			}
+		default:
+			c.lggr.Warnw("Unknown log event signature",
+				"blockNumber", log.BlockNumber, "eventSig", log.EventSig, "logHash", log.TxHash.Hex())
 			continue
 		}
-		unpacked := new(channel_config_store.ChannelConfigStoreNewChannelDefinition)
+		c.lggr.Infow("Got new logs", "source", trigger.Source, "url", trigger.URL, "sha", hex.EncodeToString(trigger.SHA[:]), "blockNum", trigger.BlockNum)
+		c.fetchTriggerCh <- trigger
+	}
+}
 
-		err := channelConfigStoreABI.UnpackIntoInterface(unpacked, newChannelDefinitionEventName, log.Data)
-		if err != nil {
-			return fmt.Errorf("failed to unpack log data: %w", err)
+// mergeDefinitions reconciles new channel definitions with the current set according to source
+// authority rules. Owner definitions (SourceOwner) have full authority: they can add, update, or
+// tombstone (delete) channels. Adder definitions (non-owner sources) have limited authority: they
+// can only add new channels and cannot overwrite or tombstone existing ones.
+//
+// Adder limits are enforced:
+//   - MaxChannelsPerAdder: The total number of channels in a single adder definition file cannot
+//     exceed this limit. This is checked before processing any channels.
+//
+// Returns an error if adder limits are exceeded
+//   - errChannelsPerAdderLimitExceeded: When the adder definition file exceeds MaxChannelsPerAdder
+func (c *channelDefinitionCache) mergeDefinitions(source uint32, currentDefinitions llotypes.ChannelDefinitions, newDefinitions llotypes.ChannelDefinitions) (llotypes.ChannelDefinitions, error) {
+	if source > SourceOwner {
+		if len(newDefinitions) > MaxChannelsPerAdder {
+			return nil, fmt.Errorf("%w: %d, max %d",
+				errChannelsPerAdderLimitExceeded, len(newDefinitions), MaxChannelsPerAdder)
 		}
-		if len(log.Topics) < 2 {
-			// should never happen but must guard against unexpected panics
-			c.lggr.Warnw("Log missing expected topics", "log", log)
-			continue
-		}
-		unpacked.DonId = new(big.Int).SetBytes(log.Topics[1])
+	}
 
-		//nolint:gosec // disable G115
-		unpacked.Raw.BlockNumber = uint64(log.BlockNumber)
+	for channelID, def := range newDefinitions {
+		switch source {
+		case SourceOwner:
+			if def.Tombstone {
+				existing, exists := currentDefinitions[channelID]
+				if exists && (existing.Source == SourceOwner && !existing.Tombstone) {
+					c.lggr.Warnw("invalid channel tombstone, cannot tombstone owner-defined channel, must be dropped from definitions",
+						"channelID", channelID, "source", source)
+					continue
+				}
+			}
```

Comment:
Following our previous discussion, we might want to remove this check to allow the owner to (simultaneously) remove a channel and tombstone it in one round (i.e., instead of removing a channel first, and later adding a tombstone). This prevents an adder from re-assigning to the same channelID in between.

Reply:
Actually, since we are now persisting the sources (with tombstones!) and the cache is independent from the plugin (it does not depend on any plugin/outcome state to update its own internal state), we might be able to avoid passing down tombstones at all to the plugin (might be worth considering, more from a cleanliness perspective than from strict performance perspective). Here, we could do:

```suggestion
			if def.Tombstone {
				delete(currentDefinitions, channelID)
			} else {
                currentDefinitions[channelID] = def
            }
```

Then, in the `Definitions` method (lines 812-851), we make sure to call `mergeDefinitions` on the `SourceOwner` *last* (see comment at `mergeDefinitions`). If the `SourceOwner` ever leaves a tombstone on a channelID, this overwrites any adder attempts to write to that channel.